pkg/executor/docker/executor.go

```diff
@@ -95,277 +110,259 @@ func (e *Executor) Run(
+			Str("job", request.JobID).
+			Logger(),
+		ID:          e.ID,
+		executionID: request.ExecutionID,
+		containerID: jobContainer.ID,
+		resultsDir:  request.ResultsDir,
+		limits:      request.OutputLimits,
+		keepStack:   config.ShouldKeepStack(),
+		waitCh:      make(chan bool),
+		activeCh:    make(chan bool),
+		running:     atomic.NewBool(false),
+	}
 
-		// create a mount so the output data does not need to be copied back to the host
-		mounts = append(mounts, mount.Mount{
+	// register the handler for this executionID
+	e.handlers.Put(request.ExecutionID, handler)
+	// run the container.
+	go handler.run(ctx)
+	return nil
+}
 
-			Type: mount.TypeBind,
-			// this is an output volume so can be written to
-			ReadOnly: false,
+// Wait waits for the completion of a specific execution by its executionID.
+// A channel is returned that will emit the result of the execution.
+// TODO: The current implementation spawns a new Goroutine for each Wait call,
+// which could become resource-intensive with a large number of concurrent Wait calls.
+// We should consider using a shared broadcast channel in the executionHandler to notify all
+// waiting Goroutines at once.
+func (e *Executor) Wait(ctx context.Context, executionID string) (<-chan *models.RunCommandResult, error) {
+	handler, found := e.handlers.Get(executionID)
+	if !found {
+		return nil, fmt.Errorf("execution (%s) not found", executionID)
+	}
+	ch := make(chan *models.RunCommandResult)
+	// TODO this pattern is wasteful if we expect there to be a large number things calling wait, as each wait spins off
+	// a go rotutine
+	go e.doWait(ctx, ch, handler)
+	return ch, nil
+}
 
-			// we create a named folder in the job results folder for this output
-			Source: srcd,
+func (e *Executor) doWait(ctx context.Context, out chan *models.RunCommandResult, handle *executionHandler) {
+	log.Info().Str("executionID", handle.executionID).Msg("waiting on execution")
+	defer close(out)
+	select {
+	case <-ctx.Done():
+		return
+	case <-handle.waitCh:
+		log.Info().Str("executionID", handle.executionID).Msg("received results from execution")
+		out <- handle.result
+	}
+}
 
-			// the path of the output volume is from the perspective of inside the container
-			Target: output.Path,
-		})
+func (e *Executor) Cancel(ctx context.Context, executionID string) error {
+	handler, found := e.handlers.Get(executionID)
+	if !found {
+		return fmt.Errorf("execution (%s) not found", executionID)
 	}
+	return handler.kill(ctx)
+}
 
-	if _, set := os.LookupEnv("SKIP_IMAGE_PULL"); !set {
-		dockerCreds := config.GetDockerCredentials()
-		if pullErr := e.client.PullImage(ctx, dockerArgs.Image, dockerCreds); pullErr != nil {
-			pullErr = errors.Wrapf(pullErr, docker.ImagePullError, dockerArgs.Image)
-			return executor.FailResult(pullErr)
-		}
+func (e *Executor) GetOutputStream(ctx context.Context, executionID string, withHistory bool, follow bool) (io.ReadCloser, error) {
```

Comment:
I'd like to avoid forcing plugin implementations to meet the requirement of implementing `GetOutputStream` as doing so can be fairly complicated and burdensome. So agreed that we should re-think this methods presence on the interface. Ideally we solve this problem once for all current and future executor implementations. 

Accessing `stderr` and `stdout` of plugin process is [fairly straightforward](https://github.com/hashicorp/go-plugin/blob/main/client.go#L183-L193) via the client config on the go-plugin process. However, since an executor plugin is responsible for executing **many** executions, muxing the streams will still be necessary, as spinning off a plugin process for each execution doesn't seem practical. If we can enforce that every execution prefixes its output with its executionID then we should be able to filter the stream on the compute side to preserve methods such as `bacalhau log`